core/Enclave/key_operation.cpp

```diff
@@ -0,0 +1,2044 @@
+        goto out;
+
+    bio = BIO_new_mem_buf(sm2_keypair, -1); // use -1 to auto compute length
+    if (bio == NULL)
+    {
+        log_d("failed to load key pem\n");
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    pkey = PEM_read_bio_PrivateKey(bio, NULL, NULL, NULL);
+    if (pkey == NULL)
+    {
+        log_d("failed to load sm2 key\n");
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    // make decryption and compute plaintext length
+    if (EVP_PKEY_set_alias_type(pkey, EVP_PKEY_SM2) != 1)
+    {
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    if (!(dctx = EVP_PKEY_CTX_new(pkey, NULL)))
+    {
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    if (EVP_PKEY_decrypt_init(dctx) != 1)
+    {
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    if (plaintext->datalen == 0)
+    {
+        size_t strLen;
+        if (EVP_PKEY_decrypt(dctx, NULL, &strLen, ciphertext->data, (size_t)ciphertext->datalen) != 1)
+        {
+            ret = SGX_ERROR_UNEXPECTED;
+            goto out;
+        }
+        plaintext->datalen = strLen;
+        ret = SGX_SUCCESS;
+        goto out;
+    }
+
+    if (ciphertext->data != NULL)
+    {
+        size_t strLen = plaintext->datalen;
+        if (EVP_PKEY_decrypt(dctx, plaintext->data, &strLen, ciphertext->data, (size_t)ciphertext->datalen) != 1)
+        {
+            ret = SGX_ERROR_UNEXPECTED;
+            goto out;
+        }
+    }
+    else
+    {
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+out:
+    BIO_free(bio);
+    EVP_PKEY_free(pkey);
+    EVP_PKEY_CTX_free(dctx);
+
+    memset_s(sm2_keypair, cmk->keybloblen, 0, cmk->keybloblen);
+    SAFE_FREE(sm2_keypair);
+
+    return ret;
+}
+
+/**
+ * @brief make rsa sign with the designated digest mode and padding mode
+ * digest mode and padding mode is optional
+ * running in enclave
+ * @param cmk_blob cipher block for storing keys
+ * @param data data to be signed
+ * @param signature used to receive signature
+ * @return sgx_status_t
+ */
+sgx_status_t ehsm_rsa_sign(const ehsm_keyblob_t *cmk,
+                           const ehsm_data_t *data,
+                           ehsm_data_t *signature)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    //verify padding mode
+    if (cmk->metadata.padding_mode != EH_PAD_RSA_PKCS1 && cmk->metadata.padding_mode != EH_PAD_RSA_PKCS1_PSS)
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    uint8_t *rsa_keypair = NULL;
+    BIO *bio = NULL;
+    RSA *rsa_prikey = NULL;
+    EVP_PKEY *evpkey = NULL;
+    EVP_MD_CTX *mdctx = NULL;
+    EVP_PKEY_CTX *pkey_ctx = NULL;
+    size_t temp_signature_size = 0;
+    // Get Digest Mode
+    const EVP_MD *digestMode = GetDigestMode(cmk->metadata.digest_mode);
+    if (digestMode == NULL)
+    {
+        log_d("ecall rsa_sign digest Mode error.\n");
+        ret = SGX_ERROR_INVALID_PARAMETER;
+        goto out;
```

Comment:
return SGX_ERROR_INVALID_PARAMETER  directly.

Reply:
it will be changed

Reply:
not updated?